### PR TITLE
Migrate to new auth flow and use Garth, fixes #103

### LINKED ISSUE #102: Improve authentication by migrating to Garth
[Garth](https://github.com/matin/garth) uses the same API as the mobile app for login and request authentication. It's as close as we can get to proper Garmin Connect API. The migration eliminates the need for Cloudscraper.

Garth ...
1. Uses OAuth1 and OAuth2 token authentication after initial login
1. OAuth1 token survive for a year
1. Supports MFA
1. Auto-refresh of OAuth2 token when expired
1. Works on Google Colab
1. Responses return JSON (instead of needing to parse HTML)

garminconnect has [already been migrated](https://github.com/cyberjunky/python-garminconnect/pull/144).

I'm happy to write the PR if you're interested in making the change.

### LINKED ISSUE #103: Breaking again - {"message":"HTTP 403 Forbidden","error":"ForbiddenException"}
Seems like it's breaking again...

garmin-get-activity --password password1 --log-level DEBUG username@gmail.com 12xxxxxxx fit           
2023-09-26 16:59:47,883 [DEBUG] using session factory: cloudscraper
2023-09-26 16:59:47,903 [INFO] authenticating user ...
2023-09-26 16:59:47,903 [INFO] fetching CSRF token ...
2023-09-26 16:59:47,937 [DEBUG] Starting new HTTPS connection (1): sso.garmin.com:443
2023-09-26 16:59:48,394 [DEBUG] https://sso.garmin.com:443 "GET /sso/login?service=https%3A%2F%2Fconnect.garmin.com%2Fmodern%2F&gauthHost=https%3A%2F%2Fsso.garmin.com%2Fsso HTTP/1.1" 200 None
2023-09-26 16:59:49,520 [DEBUG] https://sso.garmin.com:443 "POST /sso/signin?service=https%3A%2F%2Fconnect.garmin.com%2Fmodern%2F&gauthHost=https%3A%2F%2Fsso.garmin.com%2Fsso HTTP/1.1" 200 None
2023-09-26 16:59:49,521 [DEBUG] got auth response: <!DOCTYPE html>
<html class="no-js">
	<head>
		<title>Success</title>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="IE=edge;" />
		<meta name="description" content="">
		<meta name="viewport" content="width=device-width, initial-scale=1">
		<meta http-equiv="cleartype" content="on">
		<script type="text/javascript" src="/sso/js/jquery/3.1.1/jquery.min.js?20210319"></script>
		<script type="text/javascript">jQuery.noConflict();</script>
		<script type="text/javascript" src="/sso/js/json2.js"></script>
		<script type="text/javascript" src="/sso/js/consoleUtils.js?20210319"></script>
		<script type="text/javascript" src="/sso/js/postmessage.js?20210319"></script>
		<script type="text/javascript">
			var redirectAfterAccountLoginUrl 	  = "";
			var redirectAfterAccountCreationUrl = "";
			var consumeServiceTicket         	  = "true";
			var service_url                  	  = "https:\/\/connect.garmin.com\/modern\/";
			var parent_url                   	  = "";
			var response_url                 	  = "https:\/\/connect.garmin.com\/modern\/?ticket=ST-0357568-mLBwQntqJcj3ajh2biCW-cas";
			var logintoken                   	  = "";
			var socialLogin                   	  = "";
			var performMFACheck                 = "";

			// Decode url if it's encoded unnecessarily (which is happening when SSO GAuth logins redisplay the login page due to session timeouts.)
			if (response_url.indexOf('%3A%2F%2F') != -1) {
				response_url = decodeURIComponent(response_url);
			}
			response_url = response_url.replace(new RegExp("&amp;", 'g'),"&");

			var service_ticket = response_url.substring(response_url.indexOf('ticket=') + 7, response_url.length);

			if (redirectAfterAccountLoginUrl) {
				consoleInfo('casEmbedSuccess.html: redirectAfterAccountLoginUrl: [' + redirectAfterAccountLoginUrl + ']');
			}
			consoleInfo('casEmbedSuccess.html: consumeServiceTicket:         [' + consumeServiceTicket + ']');
			consoleInfo('casEmbedSuccess.html: service_url:                  [' + service_url + ']');
			consoleInfo('casEmbedSuccess.html: parent_url:                   [' + parent_url + ']');
			consoleInfo('casEmbedSuccess.html: response_url:                 [' + response_url + ']');
			consoleInfo('casEmbedSuccess.html: service_ticket:               [' + service_ticket + ']');
			if (logintoken) {
				consoleInfo('casEmbedSuccess.html: logintoken:                   [' + logintoken + ']');
			}
			if (socialLogin) {
				consoleInfo('casEmbedSuccess.html: socialLogin:                   [' + socialLogin + ']');
			}

			function send(msg) {
				consoleInfo('casEmbedSuccess.html: send(): Calling XD.postMessage(msg:[' + JSON.stringify(msg) + '], target_url:[' + parent_url + '])...');
				XD.postMessage(msg, parent_url, parent);
				return false;
			}

			function redirect(target){
				var embedWidget = "";
				if (embedWidget != 'true') {
					if (logintoken || socialLogin) {
						//  Tell parent to close opened Gauth lite box.
						send({'closeLiteBox':'1'});
					} else {
						// Tell parent to resize Gauth widget's height to just that of the loading icon.
						send({'gauthHeight':jQuery("#GAuth-component").height()});
					}
				}

				if (window.opener != null) {
					var iframeParent = window.opener.parent;
					iframeParent.location.href = target;
					window.close();
				} else {
					// If the "redirectAfterAccountLoginUrl" GAuth config parameter was specified,
					// a full page refresh upon logon was requested, so redirect the parent window upon login.
					if (redirectAfterAccountLoginUrl || redirectAfterAccountCreationUrl || (socialLogin && socialLogin === 'true') || (performMFACheck && performMFACheck === 'true')) {

						if (embedWidget != 'true') {
							send({'status':'SUCCESS', 'successDetails':'Login Successful'});
						}

						// Redirect parent of login iframe to requested service URL with ?ST=<Service Ticket ID> appended,
						// so the webapp can validate the service ticket and log the user on to the webapp.
						consoleInfo('casEmbedSuccess.html: Calling parent.location.href = [' + response_url + '];...');
						top.location.href = response_url;

					// Else if GAuth was configured not to consume the service ticket, send the service ticket and service url.
					// It's then up to the parent page to hide the GAuth widget and validate the service ticket.
					} else if (consumeServiceTicket == 'false') {

						send({'status':'SUCCESS', 'successDetails':'Login Successful', 'serviceTicket':service_ticket, 'serviceUrl':service_url});

					} else {
						// Else consume the service ticket and log the user into the parent webapp, by making a JSONP request to the response url.
						// This service ticket request to the parent webapp will return a small amount of JSON that we
						// can send in the success event to the parent page to notify them of successful logon
						// It's then up to the parent page to hide the GAuth widget and do whatever it wants.
						consoleInfo('casEmbedSuccess.html: Loading ajax jsonp URL: [' + response_url + ']');
						jQuery.ajax({
							type: "REDIRECT",
							url: response_url,
							dataType: 'jsonp',
							error: function(xhr, status, error) {
								consoleError('casEmbedSuccess.html: Error loading ajax jsonp URL: [' + response_url + ']! Error: ' + error);
							},
							success: function( data, status, xhr ) {
								consoleInfo('casEmbedSuccess.html: success loading ajax jsonp url. data: [' + data + ']');
								var userdata = data;
								if (typeof(userdata) === 'string') {
									userdata = JSON.parse(data);
								}
								consoleInfo('casGenericRedirect.jsp: customerId: ['            + userdata.customerId
																			+ '], username: ['               + userdata.username
																			+ '], rememberMe: ['             + userdata.rememberMe
																			+ '], password: ['               + userdata.password
																			+ '], email: ['                  + userdata.email
																			+ '], displayName: ['            + userdata.displayName
																			+ '], firstName: ['              + userdata.firstName
																			+ '], lastName: ['               + userdata.lastName
																			+ '], localePreference: ['       + userdata.localePreference
																			+ '], addressLine1: ['           + userdata.addressLine1
																			+ '], addressLine2: ['           + userdata.addressLine2
																			+ '], cityName: ['               + userdata.cityName
																			+ '], state: ['                  + userdata.state
																			+ '], postalCode: ['             + userdata.postalCode
																			+ '], country: ['                + userdata.country
																			+ '], passwordChangeRequired: [' + userdata.passwordChangeRequired
																			+ '], lastLogin: ['              + userdata.lastLogin
																			+ '], erpCustomerNumber: ['      + userdata.erpCustomerNumber
																			+ ']');
								send({'status'                 : 'SUCCESS',
											'successDetails'         : 'Login Successful',
											'customerId'             : userdata.customerId,
											'username'               : userdata.username,
											'rememberMe'             : userdata.rememberMe,
											'password'               : userdata.password,
											'email'                  : userdata.email,
											'displayName'            : userdata.displayName,
											'firstName'              : userdata.firstName,
											'lastName'               : userdata.lastName,
											'localePreference'       : userdata.localePreference,
											'addressLine1'           : userdata.addressLine1,
											'addressLine2'           : userdata.addressLine2,
											'cityName'               : userdata.cityName,
											'state'                  : userdata.state,
											'postalCode'             : userdata.postalCode,
											'country'                : userdata.country,
											'passwordChangeRequired' : userdata.passwordChangeRequired,
											'lastLogin'              : userdata.lastLogin,
											'erpCustomerNumber'      : userdata.erpCustomerNumber
								});
							}
						});
					}
				}
			}
		</script>
	</head>
	<body>
		<div id="GAuth-component">
			
		</div>
		<script type="text/javascript">
			jQuery(document).ready(function(){
				var service = "https:\/\/connect.garmin.com\/modern\/";
				consoleInfo("casEmbedSuccess.html: ready, calling redirect('" + service + "')...");
				redirect(service);
			});
		</script>
	<script>(function(){var js = "window['__CF$cv$params']={r:'80ca501c3a3713ea',t:'MTY5NTcxODc4OS41NzQwMDA='};_cpo=document.createElement('script');_cpo.nonce='',_cpo.src='/cdn-cgi/challenge-platform/scripts/jsd/main.js',document.getElementsByTagName('head')[0].appendChild(_cpo);";var _0xh = document.createElement('iframe');_0xh.height = 1;_0xh.width = 1;_0xh.style.position = 'absolute';_0xh.style.top = 0;_0xh.style.left = 0;_0xh.style.border = 'none';_0xh.style.visibility = 'hidden';document.body.appendChild(_0xh);function handler() {var _0xi = _0xh.contentDocument || _0xh.contentWindow.document;if (_0xi) {var _0xj = _0xi.createElement('script');_0xj.innerHTML = js;_0xi.getElementsByTagName('head')[0].appendChild(_0xj);}}if (document.readyState !== 'loading') {handler();} else if (window.addEventListener) {document.addEventListener('DOMContentLoaded', handler);} else {var prev = document.onreadystatechange || function () {};document.onreadystatechange = function (e) {prev(e);if (document.readyState !== 'loading') {document.onreadystatechange = prev;handler();}};}})();</script></body>
</html>

2023-09-26 16:59:49,522 [DEBUG] auth ticket url: 'https://connect.garmin.com/modern/?ticket=ST-0357568-mLBwQntqJcj3ajh2biCW-cas'
2023-09-26 16:59:49,522 [INFO] claiming auth ticket ...
2023-09-26 16:59:49,524 [DEBUG] Starting new HTTPS connection (1): connect.garmin.com:443
2023-09-26 16:59:50,106 [DEBUG] https://connect.garmin.com:443 "GET /modern/?ticket=ST-0357568-mLBwQntqJcj3ajh2biCW-cas HTTP/1.1" 302 0
2023-09-26 16:59:50,373 [DEBUG] https://connect.garmin.com:443 "GET /modern/ HTTP/1.1" 302 0
2023-09-26 16:59:50,636 [DEBUG] https://connect.garmin.com:443 "GET /signin/?service=https%3A%2F%2Fconnect.garmin.com%2Fmodern%2F HTTP/1.1" 200 None
2023-09-26 16:59:50,904 [DEBUG] https://connect.garmin.com:443 "GET /modern HTTP/1.1" 302 0
2023-09-26 16:59:51,260 [DEBUG] https://sso.garmin.com:443 "GET /sso/login?service=https%3A%2F%2Fconnect.garmin.com%2Fmodern%2F&webhost=https%3A%2F%2Fconnect.garmin.com&gateway=true&generateExtraServiceTicket=true&generateTwoExtraServiceTickets=true&clientId=GarminConnect HTTP/1.1" 302 0
2023-09-26 16:59:51,702 [DEBUG] https://connect.garmin.com:443 "GET /modern/ HTTP/1.1" 302 0
2023-09-26 16:59:51,974 [DEBUG] https://connect.garmin.com:443 "GET /signin/?service=https%3A%2F%2Fconnect.garmin.com%2Fmodern%2F HTTP/1.1" 200 None
2023-09-26 16:59:51,975 [INFO] fetching activity 12091532009 ...
2023-09-26 16:59:52,307 [DEBUG] https://connect.garmin.com:443 "GET /proxy/activity-service/activity/12091532009 HTTP/1.1" 403 None
2023-09-26 16:59:52,309 [ERROR] failed to fetch json summary for activity 12xxxxxxx: 403
{"message":"HTTP 403 Forbidden","error":"ForbiddenException"}
2023-09-26 16:59:52,309 [ERROR] failed with exception: failed to fetch json summary for activity 12xxxxxxx: 403
{"message":"HTTP 403 Forbidden","error":"ForbiddenException"}
Traceback (most recent call last):
  File "/Users/username/.pyenv/versions/3.11.3/bin/garmin-get-activity", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/username/.pyenv/versions/3.11.3/lib/python3.11/site-packages/garminexport/cli/get_activity.py", line 68, in main
    summary = client.get_activity_summary(args.activity)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/username/.pyenv/versions/3.11.3/lib/python3.11/site-packages/garminexport/garminclient.py", line 73, in check_session
    return client_function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/username/.pyenv/versions/3.11.3/lib/python3.11/site-packages/garminexport/garminclient.py", line 290, in get_activity_summary
    raise Exception(u"failed to fetch json summary for activity {}: {}\n{}".format(
Exception: failed to fetch json summary for activity 12xxxxxxx: 403
{"message":"HTTP 403 Forbidden","error":"ForbiddenException"}

### PR DESCRIPTION
**Pull request by SimonBaars**
For full context, see the relevant issue thread: https://github.com/pe-st/garmin-connect-export/issues/95

TLDR: Garmin shut down the old authentication flow. The new authentication flow has many security measures in place to script the login. The solution described in issue #102 works and we can use the Garth library to implement an OAuth flow, which should be more robust than the original method.

The OAuth flow also required using the new endpoints (which could be migrated by simply removing the proxy).

Thanks to @geraudloup for investigating this with me, and @app4g for the support!

### CODE REVIEW

**Comment by matin** [comment]
@SimonBaars love this!

Why not use `garth.connectapi()` for the requests though?

`garth.connectapi()` routes to connectapi.garmin.com vs connect.garmin.com (and rely on setting the backend). The method also handles auto-renew of the OAuth2 Bearer token and special cases like a 204 response.

Take a look at how Garth is being used in [garminconnect](https://github.com/cyberjunky/python-garminconnect/blob/master/garminconnect/__init__.py)

I'm the maintainer of Garth. Let me know how I can be helpful.

**Comment by SimonBaars** [comment]
Hey @matin, good to have you around! FYI, I implemented Garth in a couple of affected repos:
- https://github.com/pe-st/garmin-connect-export/pull/96
- https://github.com/petergardfjall/garminexport/pull/104
- https://github.com/jaroslawhartman/withings-sync/pull/129

I agree that using those Garth features as well would be better, but for now I wanted to stick as close as possible to the original implementation to not make it a too great diff at once (and also not too dependent on Garth, in case any of the maintainers would prefer a different direction later). We can always make it a separate effort 😄

**Comment by SimonBaars** [comment]
I changed the PR title to show that the main goal is to fix auth.

**Comment by matin** [comment]
Fair point, and thanks for the work on the migrations!

**Comment by petergardfjall** [comment]
Thanks for the contribution. I need to understand the implications better though.
In particular, https://github.com/petergardfjall/garminexport/issues/102#issuecomment-1742000911.

**Comment by matin** [comment]
> Thanks for the contribution. I need to understand the implications better though. In particular, [#102 (comment)](https://github.com/petergardfjall/garminexport/issues/102#issuecomment-1742000911).

Responded in #102

**Comment by petergardfjall** [comment]
> Garmin shut down the old authentication flow.

The "old authentication flow" (the one used by the web browser) is not shut down, but Garmin Connect uses Cloudflare's bot protection to prevent scripted access. That's why trying to replicate browser requests with `curl` gives a `403 (Forbidden)` response. Cloudflare recognizes (by the TLS handshake fingerprint) that the request did not come from a web browser.

One solution that resolves the present [403 (Forbidden) issue](https://github.com/petergardfjall/garminexport/issues/103) on my end is to use [curl_cffi](https://github.com/yifeikong/curl_cffi), a `requests` replacement that relies on https://github.com/lwthiker/curl-impersonate to perform requests that Cloudflare believes are coming from a web browser. 

With that, the "old" (browser-based) authentication flow that `garminexport` has always relied on seems to work fine.

**Comment by app4g** [comment]
@SimonBaars if you look at my comment at https://github.com/pe-st/garmin-connect-export/issues/95#issuecomment-1736800945, when I login manually via chrome, I Also did get a SESSIONID cookie, but it was further down the flow rather than earlier.

**Comment by petergardfjall** [comment]
For reasons stated in [here](https://github.com/petergardfjall/garminexport/issues/102#issuecomment-1806817542) I will not merge this PR.

Appreciate the contribution and discussion though.
